### PR TITLE
Auxiliary Image Domain (V8 Schema) Upgrade 

### PR DESCRIPTION
**Pull request by anpanigr**
New Auxiliary Image Domain ( with v8 schema )  Upgrade Test 
Note : These tests tests are currently commented out. It will be uncommentated once WKO code to aux upgrade is merger to main.  

  a.  ItAuxDomainImplicitUpgrde
          Start a Aux domain with v8 schema in main branch.  Make sure the Operator with webhook convert the Domain 
          resource to k8s based artifacts and start the domain.
  b. ItOperatorWlsUpgrade
        Start a Aux domain with v8 schema  in WKO v3.3.8.  Upgrade the Operator to latest
        Make sure the Domain schema is updated  with k8s based artifacts 

kind new results 
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/9386
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/9395

### CODE REVIEW

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    List<String> archiveList = Collections.singletonList(ARCHIVE_DIR + "/" + MII_BASIC_APP_NAME + ".zip");
+
+    List<String> modelList = new ArrayList<>();
+    modelList.add(MODEL_DIR + "/" + MII_BASIC_WDT_MODEL_FILE);
+    modelList.add(MODEL_DIR + "/model.jms2.yaml");
+
+    // create auxiliary image using imagetool command if does not exists
+    logger.info("creating auxiliary image {0}:{1} using imagetool.sh ", miiAuxiliaryImage, MII_BASIC_IMAGE_TAG);
+    createPushAuxiliaryImageWithDomainConfig(miiAuxiliaryImage, archiveList, modelList);
+
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java:144]: add a comment here about generating the domain.yaml

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    StringBuffer getDomain = new StringBuffer("kubectl get domain ");
+    getDomain.append(domainUid);
+    getDomain.append(" -n " + domainNamespace);
+    getDomain.append(" -o yaml | grep compatibility-mode-operator");
+    // Get the domain in yaml format
+    ExecResult dresult = assertDoesNotThrow(
+                 () -> exec(new String(getDomain), true));
+    logger.info("Get domain command {0}", getDomain.toString());
+    logger.info("kubectl get domain returned {0}", dresult.toString());
+    assertTrue(dresult.stdout().contains("compatibility-mode-operator"), "Failed to implicitly upgrade v8 aux domain");
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java:197]: @rjeberhard @ankedia do we need this check or checking for domain successfully created is enough?

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java [hunk 3]
```diff
@@ -179,6 +203,75 @@ void testOperatorWlsUpgradeFrom338ToLatest(String domainType) {
...
+  // once the code to upgrade v8 style auximage domain is merged into main branch 
+  @Disabled()
+  @Test
+  @DisplayName("Upgrade 3.3.8 Domain(v8 schema) with Auxiliary Image to latest")
+  void testOperatorWlsAuxDomainUpgradeFrom338ToLatest() {
+    logger.info("Starting test to upgrade Domain with Auxiliary Image with v8 schema to latest ");
+
+    installOldOperator("3.3.8");
+    createSecrets();
+
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java:219]: add comment creating aux image here

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java [hunk 4]
```diff
@@ -179,6 +203,75 @@ void testOperatorWlsUpgradeFrom338ToLatest(String domainType) {
...
+    createSecrets();
+   
+    final String auxiliaryImagePath = "/auxiliary";
+    List<String> archiveList = Collections.singletonList(ARCHIVE_DIR + "/" + MII_BASIC_APP_NAME + ".zip");
+    List<String> modelList = new ArrayList<>();
+    modelList.add(MODEL_DIR + "/" + MII_BASIC_WDT_MODEL_FILE);
+    modelList.add(MODEL_DIR + "/model.jms2.yaml");
+    logger.info("creating auxiliary image {0}:{1} using imagetool.sh ", miiAuxiliaryImage, MII_BASIC_IMAGE_TAG);
+    createPushAuxiliaryImageWithDomainConfig(miiAuxiliaryImage, archiveList, modelList);
+
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java:227]: add comment generating domain yaml

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java [hunk 5]
```diff
@@ -179,6 +203,75 @@ void testOperatorWlsUpgradeFrom338ToLatest(String domainType) {
...
+    checkDomainStarted(domainUid, domainNamespace);
+    LinkedHashMap<String, OffsetDateTime> pods = new LinkedHashMap<>();
+    pods.put(adminServerPodName, getPodCreationTime(domainNamespace, adminServerPodName));
+    // get the creation time of the managed server pods before upgrading the operator
+    for (int i = 1; i <= replicaCount; i++) {
+      pods.put(managedServerPodNamePrefix + i, getPodCreationTime(domainNamespace, managedServerPodNamePrefix + i));
+    }
+
+    // verify there is no status condition type Completed before upgrading to Latest
+    verifyDomainStatusConditionTypeDoesNotExist(domainUid, domainNamespace,
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java:267]: do we need to check domain status after upgrade?

Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java:267]: I just took the same code from the existing test method.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java [hunk 6]
```diff
@@ -196,34 +289,105 @@ public void tearDown() {
...
-
-    latestOperatorImageName = getOperatorImageName();
+    domainNamespace = namespaces.get(1);
 
     // install operator with older release 
     HelmParams opHelmParams = installOperator(operatorVersion, 
                  opNamespace, domainNamespace);
+  }
+
+  // After upgrdae scale up/down the cluster
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java:305]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java (hunk 1 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java:144]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java (hunk 3 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java:219]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java (hunk 4 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java:228]: done

**Review by vanajamukkara — APPROVED** [review]
LG

**Comment by ankedia** [comment]
General questions for @anpanigr , @rjeberhard and @tbarnes-us -
1) Do we need tests for multiple auxiliary images using 3.x auxiliary image feature defined under `ServerPod`?
2) Do we need tests for cluster scoped and server scoped 3.x style auxiliary images?
3) Do we need tests that define multiple 3.x style auxiliary image volumes?

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java (hunk 2 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java:197]: This check is th only way to know the implicit upgrade has happened.

Comment (rjeberhard) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java:197]: It's odd to do a contains()... Can you at least look for the specific field you want in the `kubectl get`?

**Comment by rjeberhard** [comment]
> General questions for @anpanigr , @rjeberhard and @tbarnes-us -

@ankedia, what's your opinion on these questions? We don't have to test every case with integration tests, so what's most likely to break that would be worth testing?

**Comment by ankedia** [comment]
> @ankedia, what's your opinion on these questions? We don't have to test every case with integration tests, so what's most likely to break that would be worth testing?
I don't have a strong opinion, we do have unit tests that covers different cases. @anpanigr discussed this in the scrum meeting with @tbarnes-us and me. Tom suggested to keep all the existing 3.x auxiliary image integration tests to verify that automated upgrade of 3.x AI works for all different scenarios.

**Comment by rjeberhard** [comment]
> keep all the existing 3.x auxiliary image integration tests to verify that automated upgrade of 3.x AI works for all different scenarios.

Just making sure that we're on the same page. I think this means that the integration tests would need to preserve a copy of the v8 model classes so that it could create a v8 version with the 3.x style AI configuration.

**Comment by ankedia** [comment]
> Just making sure that we're on the same page. I think this means that the integration tests would need to preserve a copy of the v8 model classes so that it could create a v8 version with the 3.x style AI configuration.
Yes, either that or use snake yaml to generate the v8 version of domain yaml using config maps (similar to the unit tests).

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java (hunk 2 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java:197]: May be when all the servers are up and runtime application is verified,  it is not necessary to verify if the implicit schema upgrade has happened ?

**Comment by anpanigr** [comment]
> > Just making sure that we're on the same page. I think this means that the integration tests would need to preserve a copy of the v8 model classes so that it could create a v8 version with the 3.x style AI configuration.
> > Yes, either that or use snake yaml to generate the v8 version of domain yaml using maps (similar to the unit tests).

We will verify the multiple aux images with different (domain, cluster) scope and make sure that all images merge order is preserved as per public doc. The negative tests case (missing WDT binary, missing DomainConfig , FailCustomMountCommand ) are of low priority.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java (hunk 2 above)
Comment (rjeberhard) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java:197]: I think that we should verify that the schema is updated.

Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxDomainImplicitUpgrde.java]: Modified the tests to verify the InitContainer name in Spec/ServerPod section of domain resource schema

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java (hunk 5 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorWlsUpgrade.java:268]: Removed the redundant check

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxV8DomainImplicitUpgrade.java [hunk 7]
```diff
@@ -0,0 +1,623 @@
...
+  private static String domainNamespace = null;
+  private static LoggingFacade logger = null;
+  private final int replicaCount = 2;
+  private static String adminSecretName;
+  private static String encryptionSecretName;
+  private static Map<String, OffsetDateTime> podsWithTimeStamps = null;
+  private boolean foundCompatiblityContainer = false;
+
+  /**
+   * Install Operator.
```
Comment (ankedia) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxV8DomainImplicitUpgrade.java:88]: Should the comment say something like `Test Initialization` as it seems to be doing more than installing operator.

**Review by ankedia — APPROVED** [review]
LGTM. I left couple of minor comments.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxV8DomainImplicitUpgrade.java (hunk 7 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItAuxV8DomainImplicitUpgrade.java:88]: @DisplayName("Test implicit upgrade of auximage domain resource with api version v8")

**Pull request merged**